File: archive/lc_1483. Kth Ancestor of a Tree Node.py
class TreeAncestor:

    def __init__(self, n: int, parent):
        pas = [[pa] for pa in parent]
        for step in itertools.count(1):
            flag = False
            for i in range(1, n):
                d = len(pas[i])
                if d == step and len(pas[pas[i][-1]]) >= d and pas[pas[i][-1]][d-1] != -1:
                    flag = True
                    pas[i].append(pas[pas[i][-1]][d-1])
            if not flag: break
        self.pas = pas

    def getKthAncestor(self, node: int, k: int) -> int:
        while k and node != -1:
            d = min(len(self.pas[node])-1, int(math.log2(k)))
            node = self.pas[node][d]
            k -= 1 << d
        return node


# Your TreeAncestor object will be instantiated and called as such:
# obj = TreeAncestor(n, parent)
# param_1 = obj.getKthAncestor(node,k)

File: archive/lc_1483. Kth Ancestor of a Tree Node.cpp
// DP, Binary Lifting, 倍增

const int K = 16;
array<array<int, K>, 50000> anc;

class TreeAncestor {
public:
    TreeAncestor(int n, vector<int>& parent) {
        for (int i=0; i<n; ++i)
            anc[i][0] = parent[i];
        for (int k=1; k<K; ++k)
            for (int i=0; i<n; ++i)
                anc[i][k] = anc[i][k-1] != -1 ? anc[anc[i][k-1]][k-1] : -1;
    }
    
    int getKthAncestor(int node, int k) {
        for (int i=0; i<K && node!=-1; ++i)
            if (k & 1<<i) node = anc[node][i];
        return node;
    }
};

/**
 * Your TreeAncestor object will be instantiated and called as such:
 * TreeAncestor* obj = new TreeAncestor(n, parent);
 * int param_1 = obj->getKthAncestor(node,k);
 */